Firestore/core/src/firebase/firestore/util/string_format.h

```diff
@@ -0,0 +1,132 @@
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+namespace internal {
+
+std::string StringFormatPieces(const char* format,
+                               std::initializer_list<absl::string_view> pieces);
+
+/**
+ * Explicit ranking for formatting choices. Only useful as an implementation
+ * detail of `FormatArg`.
+ */
+template <int I>
+struct FormatChoice : FormatChoice<I + 1> {};
+
+template <>
+struct FormatChoice<3> {};
+
+}  // namespace internal
+
+/**
+ * Acts as the main value parameter to StringFormat and related functions.
+ *
+ * Chooses a conversion to a text form in this order:
+ *   * If the value is exactly of `bool` type (without implicit conversions)
+ *     the text will the "true" or "false".
+ *   * If the value is of type `const char*`, the text will be the value
+ *     interpreted as a C string.
+ *   * If the value is any other pointer type, the text will be the hexidecimal
+ *     formatting of the value as an unsigned integer.
+ *   * Otherwise the value is interpreted as anything absl::AlphaNum accepts.
+ */
+class FormatArg : public absl::AlphaNum {
+ public:
+  template <typename T>
+  FormatArg(T&& value)  // NOLINT(runtime/explicit)
+      : FormatArg{std::forward<T>(value), internal::FormatChoice<0>{}} {
+  }
+
+ private:
+  /**
+   * Creates a FormatArg from a boolean value, representing the string
+   * "true" or "false".
+   *
+   * This overload only applies if the type of the argument is exactly `bool`.
+   * No implicit conversions to bool are accepted.
+   */
+  template <typename T,
+            typename = typename std::enable_if<std::is_same<bool, T>{}>::type>
+  FormatArg(T bool_value, internal::FormatChoice<0>)
+      : AlphaNum{bool_value ? "true" : "false"} {
+  }
+
+  /**
+   * Creates a FormatArg from a character string literal. This is
+   * handled specially to avoid ambiguity with generic pointers, which are
+   * handled differently.
+   */
+  FormatArg(const char* string_value, internal::FormatChoice<1>)
+      : AlphaNum{string_value} {
+  }
+
+  /**
+   * Creates a FormatArg from an arbitrary pointer, represented as a
+   * hexidecimal integer literal.
+   */
+  template <typename T>
+  FormatArg(T* pointer_value, internal::FormatChoice<2>)
+      : AlphaNum{absl::Hex{reinterpret_cast<uintptr_t>(pointer_value)}} {
+  }
+
+  /**
+   * As a final fallback, creates a FormatArg from any type of value that
+   * absl::AlphaNum accepts.
+   */
+  template <typename T>
+  FormatArg(T&& value, internal::FormatChoice<3>)
+      : AlphaNum{std::forward<T>(value)} {
+  }
+};
+
+/**
+ * Formats a string using a simplified printf-like formatting mechanism that
+ * does not rely on C varargs.
+ *
+ * The following format specifiers are recognized:
+ *   * "%%" - A literal "%"
+ *   * "%s" - The next parameter is copied through
+ */
+template <typename... FA>
+std::string StringFormat(const char* format, const FA&... args) {
+  return internal::StringFormatPieces(
```

Comment:
I'm loathe to assert anything in here since this is intended to be part of the implementation of assert. Additionally, that would creates a circularity that maybe I could resolve by removing the `firebase_firestore_util_base` target. This seems like more trouble than it's worth. WDYT?

Reply:
Yeah, I didn't consider the circularity. One option is to silently return, but it could mask bugs. I suppose the best option is to leave as is.